### PR TITLE
feat: Convert rego queries into SQL clauses

### PR DESCRIPTION
**Pull request by Emyrk**
# What this does

Currently `Filter` works by querying all resources of a given type, then calling `Authorize()` on each element in the list. This would not work with pagination.

This feature takes the queries resulting from partial execution, and returns a SQL clause that can be added into a `WHERE` statement of a SQL query. The results are equivalent to all authorized objects in the original `Filter` call.

**This PR only implements this for listing workspaces.** This technique can be copied easily to other listing (templates, users, etc). I just started with the one with the largest lists and the one we will likely paginate first.

# Example

Partial execution returns these queries. If either is true, the user in question can read the object.

```
+---------+--------------------------------------------------------------------+
| Query 1 | input.object.org_owner != ""                                       |
|         | input.object.org_owner in {"19a3d781-9e53-4608-948e-ac084470cd99"} |
|         | input.object.owner != ""                                           |
|         | "me" = input.object.owner                                          |
+---------+--------------------------------------------------------------------+
| Query 2 | input.object.org_owner = ""                                        |
|         | input.object.owner != ""                                           |
|         | "me" = input.object.owner                                          |
+---------+--------------------------------------------------------------------+
```

This code converts these rego queries into the following SQL:

```SQL
SELECT * FROM workspaces WHERE
-- SQL generated below this line
(
    (
        input.object.org_owner != '' AND 
        input.object.org_owner = ANY(ARRAY ['a207e4ff-2cf4-44be-aadd-8f866452e3b1']) AND 
        input.object.owner != '' AND 'me' = input.object.owner
    ) 
        OR
    (
        input.object.org_owner = '' AND 
        input.object.owner != '' AND 
        'me' = input.object.owner
    )
)
```

# Long Rego Query Example

```
+---------+--------------------------------------------------------------------+
| Query 1 | input.object.org_owner != ""                                       |
|         | input.object.org_owner in {"19a3d781-9e53-4608-948e-ac084470cd99"} |
|         | input.object.owner != ""                                           |
|         | "me" = input.object.owner                                          |
+---------+--------------------------------------------------------------------+
| Query 2 | input.object.org_owner = ""                                        |
|         | input.object.owner != ""                                           |
|         | "me" = input.object.owner                                          |
+---------+--------------------------------------------------------------------+
| Query 3 | input.object.org_owner != ""                                       |
|         | input.object.org_owner in {"19a3d781-9e53-4608-948e-ac084470cd99"} |
|         | input.object.org_owner != ""                                       |
|         | "read" in input.object.acl_group_list.allUsers                     |
+---------+--------------------------------------------------------------------+
| Query 4 | input.object.org_owner != ""                                       |
|         | input.object.org_owner in {"19a3d781-9e53-4608-948e-ac084470cd99"} |
|         | input.object.org_owner != ""                                       |
|         | "*" in input.object.acl_group_list.allUsers                        |
+---------+--------------------------------------------------------------------+
| Query 5 | input.object.org_owner = ""                                        |
|         | input.object.org_owner != ""                                       |
|         | "read" in input.object.acl_group_list.allUsers                     |
+---------+--------------------------------------------------------------------+
| Query 6 | input.object.org_owner = ""                                        |
|         | input.object.org_owner != ""                                       |
|         | "*" in input.object.acl_group_list.allUsers                        |
+---------+--------------------------------------------------------------------+
| Query 7 | "read" in input.object.acl_user_list.me                            |
+---------+--------------------------------------------------------------------+
| Query 8 | "*" in input.object.acl_user_list.me                               |
**+---------+--------------------------------------------------------------------+**
```

# Future Work

Implement this technique for other list calls.

### CODE REVIEW

File: coderd/database/custom_queries.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+	"golang.org/x/xerrors"
+
+	"github.com/coder/coder/coderd/rbac"
+
+	"github.com/lib/pq"
+)
+
+// AuthorizedGetWorkspaces returns all workspaces that the user is authorized to access.
+func (q *sqlQuerier) AuthorizedGetWorkspaces(ctx context.Context, arg GetWorkspacesParams, authorizedFilter rbac.AuthorizeFilter) ([]Workspace, error) {
```
Comment (Emyrk) [inline coderd/database/custom_queries.go:15]: I wish we could add hooks into SQLc to do this. Kinda sucks to have to copy the original function and call it like this.

File: coderd/database/custom_queries.go [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+)
+
+type customQuerier interface {
+	AuthorizedGetWorkspaces(ctx context.Context, arg GetWorkspacesParams, authorizedFilter rbac.AuthorizeFilter) ([]Workspace, error)
+}
+
+// AuthorizedGetWorkspaces returns all workspaces that the user is authorized to access.
+// This code is copied from `GetWorkspaces` and adds the authorized filter WHERE
+// clause.
+func (q *sqlQuerier) AuthorizedGetWorkspaces(ctx context.Context, arg GetWorkspacesParams, authorizedFilter rbac.AuthorizeFilter) ([]Workspace, error) {
```
Comment (Emyrk) [inline coderd/database/custom_queries.go:20]: @f0ssel @coadler Will this query still be traced in our metrics? (eg datadog)

**Review by johnstcn — COMMENTED** [review]
This looks good to me, but I'll defer to other folks who are working on v2.

File: coderd/database/custom_queries.go (hunk 2 above)
Comment (coadler) [reply coderd/database/custom_queries.go:20]: Yeah these will still show up. I'd recommend adding a `-- name: QueryName :many` as the first line in the query so we can filter for it!

File: coderd/database/custom_queries.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
```
Comment (kylecarbs) [inline coderd/database/custom_queries.go:20]: I might change the wording of this. Prefixing with `Authorized` makes me think that I'm authorized to get workspaces, not that the workspaces returned I'm authorized to access.

Thoughts on `GetAuthorizedWorkspaces` instead?

Comment (kylecarbs) [reply coderd/database/custom_queries.go:20]: Is it possible for us to get rid of `GetWorkspaces` in its entirety? There doesn't seem much of a point in getting workspaces that the user doesn't have access to.

**Review by kylecarbs — COMMENTED** [review]
These tests are surprisingly simple to follow, or at least much simpler than I would have expected interfacing with Rego to be.

My primary concern is the query separation between generated and non-generated. I'd love if we could have a single query (maybe even for just getting a single?) that used the same authorization flow, but maybe that's just unreasonable.

eg.
- `GetWorkspaceByID`
- `GetWorkspaces`
- `GetWorkspaceByOwnerAndName`

Could all be replaced by this?

**Comment by Emyrk** [comment]
> AuthorizedGetWorkspaces

This change is only made for "listing" objects, for singular objects running the rego on the returned object is easier. So if we implemented this for the singular too, I would not convert to SQL, I would just call the sqlc code and run `Authorize()` on that returned object.

I wish there was a way to integrate this into SQLc, as it is a great tool, it just doesn't support this use case.

----

My goal for this PR was to implement this functionality to make this possible. We next implement it for the 2/3 list queries that are "slow" with the `Filter`. It is really only needed in 2/3 places to support that dashboard "List all that I can view" view.

**Review by kylecarbs — APPROVED** [review]

**Pull request merged**